File: results/dnd-rest-benchmark/lifecycle-runs/20260804T072004Z_claude_claude-sonnet-5_ruby-sinatra/lib/analytics.rb
# Deterministic campaign analytics: readiness scoring and risk signals,
# aggregated fresh from SQLite on every request (no caching, no randomness).

def campaign_analytics_counts(campaign_id)
  open_quests = db.execute(
    "SELECT COUNT(*) AS cnt FROM campaign_quests WHERE campaign_id = ? AND status != 'completed'",
    [campaign_id]
  ).first['cnt']

  friendly_npcs = db.execute(
    'SELECT COUNT(*) AS cnt FROM campaign_npcs WHERE campaign_id = ? AND disposition > 0',
    [campaign_id]
  ).first['cnt']

  scheduled_sessions = db.execute(
    'SELECT COUNT(*) AS cnt FROM campaign_sessions WHERE campaign_id = ?',
    [campaign_id]
  ).first['cnt']

  inventory_items = db.execute(
    'SELECT COUNT(DISTINCT item_slug) AS cnt FROM campaign_inventory WHERE campaign_id = ?',
    [campaign_id]
  ).first['cnt']

  active_quests = db.execute(
    "SELECT COUNT(*) AS cnt FROM campaign_quests WHERE campaign_id = ? AND status = 'active'",
    [campaign_id]
  ).first['cnt']

  characters = db.execute(
    'SELECT COUNT(*) AS cnt FROM campaign_characters WHERE campaign_id = ?',
    [campaign_id]
  ).first['cnt']

  {
    open_quests: open_quests,
    friendly_npcs: friendly_npcs,
    scheduled_sessions: scheduled_sessions,
    inventory_items: inventory_items,
    active_quests: active_quests,
    characters: characters
  }
end

def campaign_analytics_signals(campaign, counts)
  {
    has_dm: campaign['dm'].is_a?(String) && !campaign['dm'].empty?,
    has_characters: counts[:characters] > 0,
    has_next_session: counts[:scheduled_sessions] > 0,
    has_active_quest: counts[:active_quests] > 0
  }
end

File: results/dnd-rest-benchmark/lifecycle-runs/20260731T062307Z_claude_claude-sonnet-5_typescript-node/src/routes/analytics.ts
// Deterministic campaign analytics: a readiness score/summary and a
// maintenance risk report, both aggregated over existing campaign tables.
// Read-only — no new persistent state.
import type { ServerResponse } from "node:http";
import { db } from "../db.js";
import { sendJson } from "../http.js";
import { isPlainObject } from "../validation.js";
import { hasCampaign } from "./campaigns.js";

function countRows(table: string, campaignId: string, extraWhere = ""): number {
  const row = db
    .prepare(`SELECT COUNT(*) AS count FROM ${table} WHERE campaign_id = ?${extraWhere}`)
    .get(campaignId) as { count: number };
  return row.count;
}

function countActiveQuests(campaignId: string): number {
  return countRows("campaign_quests", campaignId, " AND status = 'active'");
}

function countFriendlyNpcs(campaignId: string): number {
  return countRows("campaign_npcs", campaignId, " AND disposition > 0");
}

function countScheduledSessions(campaignId: string): number {
  return countRows("campaign_sessions", campaignId);
}

function countInventoryItems(campaignId: string): number {
  return countRows("campaign_inventory", campaignId, " AND countable = 1");
}

function hasCampaignCharacters(campaignId: string): boolean {
  return countRows("campaign_characters", campaignId) > 0;
}

function getCampaignDm(campaignId: string): string {
  const row = db.prepare("SELECT dm FROM campaigns WHERE id = ?").get(campaignId) as { dm: string };
  return row.dm;
}

function computeReadinessScore(
  openQuests: number,
  friendlyNpcs: number,
  scheduledSessions: number,
  inventoryItems: number,
): number {
  return Math.min(100, 35 + 10 * openQuests + 15 * friendlyNpcs + 20 * scheduledSessions + 5 * inventoryItems);
}

export function handleAnalyticsSummary(res: ServerResponse, campaignId: string): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  const openQuests = countActiveQuests(campaignId);
  const friendlyNpcs = countFriendlyNpcs(campaignId);
  const scheduledSessions = countScheduledSessions(campaignId);
  const inventoryItems = countInventoryItems(campaignId);

  sendJson(res, 200, {
    campaign_id: campaignId,
    readiness_score: computeReadinessScore(openQuests, friendlyNpcs, scheduledSessions, inventoryItems),
    open_quests: openQuests,
    friendly_npcs: friendlyNpcs,
    scheduled_sessions: scheduledSessions,
    inventory_items: inventoryItems,
  });
}

export function handleRiskReport(res: ServerResponse, campaignId: string, body: unknown): void {
  if (!hasCampaign(campaignId)) {
    sendJson(res, 404, { error: "campaign not found" });
    return;
  }

  if (!isPlainObject(body) || (body.include_zeroes !== undefined && typeof body.include_zeroes !== "boolean")) {
    sendJson(res, 400, { error: "invalid request" });
    return;
  }

  const hasDm = getCampaignDm(campaignId).length > 0;
  const hasCharacters = hasCampaignCharacters(campaignId);
  const hasNextSession = countScheduledSessions(campaignId) > 0;
  const hasActiveQuest = countActiveQuests(campaignId) > 0;

  const missing: string[] = [];
  if (!hasDm) missing.push("dm");
  if (!hasCharacters) missing.push("characters");
  if (!hasNextSession) missing.push("next_session");
  if (!hasActiveQuest) missing.push("active_quest");

  const riskLevel = missing.length === 0 ? "low" : missing.length <= 2 ? "medium" : "high";

  sendJson(res, 200, {
    campaign_id: campaignId,
    risk_level: riskLevel,
    missing,
    signals: {
      has_dm: hasDm,
      has_characters: hasCharacters,
      has_next_session: hasNextSession,
      has_active_quest: hasActiveQuest,
    },
  });
}
